osu.Game.Tests/Database/LegacyExporterTest.cs

```diff
@@ -0,0 +1,122 @@
+        private TemporaryNativeStorage storage = null!;
+
+        private const string short_filename = "normal file name";
+
+        private const string long_filename =
+            "some file with super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name super long name";
+
+        [SetUp]
+        public void SetUp()
+        {
+            storage = new TemporaryNativeStorage("export-storage");
+            legacyExporter = new TestLegacyExporter(storage);
+        }
+
+        [Test]
+        public void ExportFileWithNormalNameTest()
+        {
+            var item = new TestPathInfo(short_filename);
+
+            Assert.That(item.Filename.Length, Is.LessThan(TestLegacyExporter.MAX_FILENAME_LENGTH));
+
+            exportItemAndAssert(item, short_filename);
+        }
+
+        [Test]
+        public void ExportFileWithNormalNameMultipleTimesTest()
+        {
+            var item = new TestPathInfo(short_filename);
+
+            Assert.That(item.Filename.Length, Is.LessThan(TestLegacyExporter.MAX_FILENAME_LENGTH));
+
+            //Export multiple times
+            for (int i = 0; i < 10; i++)
+            {
+                string expectedFileName = i == 0 ? short_filename : $"{short_filename} ({i})";
+                exportItemAndAssert(item, expectedFileName);
+            }
+        }
+
+        [Test]
+        public void ExportFileWithSuperLongNameTest()
+        {
+            int expectedLength = TestLegacyExporter.MAX_FILENAME_LENGTH - (legacyExporter.GetExtension().Length);
+            string expectedName = long_filename.Remove(expectedLength);
+
+            var item = new TestPathInfo(long_filename);
+
+            Assert.That(item.Filename.Length, Is.GreaterThan(TestLegacyExporter.MAX_FILENAME_LENGTH));
+            exportItemAndAssert(item, expectedName);
+        }
+
+        [Test]
+        public void ExportFileWithSuperLongNameMultipleTimesTest()
+        {
+            int expectedLength = TestLegacyExporter.MAX_FILENAME_LENGTH - (legacyExporter.GetExtension().Length);
+            string expectedName = long_filename.Remove(expectedLength);
+
+            var item = new TestPathInfo(long_filename);
+
+            Assert.That(item.Filename.Length, Is.GreaterThan(TestLegacyExporter.MAX_FILENAME_LENGTH));
+
+            //Export multiple times
+            for (int i = 0; i < 10; i++)
+                exportItemAndAssert(item, expectedName);
```

Comment:
As far as I can tell, this test is blatantly testing **broken** behaviour. Every export is overwriting the previous one. And your test is calling that a "success".

Reply:
This test is quite a problem one. As I said, `NamingUtils.GetNextBestFilename` will add 4 (and so on) characters to the end of the path which we would need to truncate again. This version truncates it after filename calculation.

Let's say `NamingUtils.GetNextBestFilename` returns "...superlongname (1)" now we need to truncate it. We can ether 
(1) truncate the name "...superlong (1)" or
(2) truncate the end: "...superlongname" (this is what I use in this PR).

If we decided to use first one, then first copy will be fine, but second one will not. Why? After getting best name first time we truncate it to "...superlong (1)". Then we run export second time. The `NamingUtils.GetNextBestFilename("...superlongname")` will look up for  "...superlongname (1)" and... wont find it, will return "...superlongname (1)" which we truncate to "...superlong (1)" and replace the existing one with this.

It can be solved by running recursion in `NamingUtils.GetNextBestFilename` which is not what I would like to do because `NamingUtils.GetNextBestFilename` definitely used somewhere else in the code.

Reply:
I still think this is wrong and needs to be addressed.

The truncation should happen in a stable way. If this means reducing the allowed length by 5 (to account for ` (99)` suffix) then that would be preferrable.